tests/jupyterhub_api/test_create_session.py

```diff
@@ -0,0 +1,162 @@
+"""
+
+"""
+To run the test run the following command:
+
+ python3 test_create_session.py --port 443 --users user2 --path /srv/jupyterhub --json '{"LCG-rel": "LCG_95a", "platform": "x86_64-centos7-gcc7-opt", "scriptenv": "none", "ncores": 2, "memory": 8589934592, "spark-cluster": "none"}'
+
+
+For multiple users
+
+python3 test_create_session.py --port 443 --users user0 user1 user2 --path /srv/jupyterhub 
+--token {token value} --json '{"LCG-rel": "LCG_95a", "platform": "x86_64-centos7-gcc7-opt", "scriptenv": "none", "ncores": 2, "memory": 8589934592, "spark-cluster": "none"}'
+"""
+
+
+import json
+import requests
+import subprocess
+import os
+import argparse
+import sys
+sys.path.append("..")
+from logger import Logger, LOG_FOLDER, LOG_EXTENSION
+from SessionUtils import Session
+from requests.packages.urllib3.exceptions import InsecureRequestWarning
+requests.packages.urllib3.disable_warnings(InsecureRequestWarning)
+import json
+import time
+
+
+def get_args():
+    parser = argparse.ArgumentParser(description='Arguments', formatter_class = argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument("--port", dest="port", type=int,
+                        required = True,
+                        )
+
+    parser.add_argument( "--users", nargs='+', dest="users",
+                        required  = True,
+                        help='list of users')
+    parser.add_argument("--delay", dest="delay",
+                        required=True,
+                        help='delay while creating the server')
+
+    parser.add_argument("--json", dest="json",
+                         required=True,
+                         help = 'json data')
+    args = parser.parse_args()
+    return args
+
+
+class CreateSession:
+    def __init__(self, port, users, params, delay, verify):
+        self.port = port
+        self.users = users
+        self.main_url = "https://localhost:" + str(self.port) + "/hub/api/"
+        self.ref_test_name = "Session_creation_test"
+        self.exit = 0
+        self.verify = verify
+        self.delay = delay
+        self.data = params
+        self.ref_timestamp = int(time.time())
+        self.session = Session()
+        self.token = self.session.get_tokens()
+        self.logger_folder = os.path.join(os.getcwd(), LOG_FOLDER)
+        self.log = Logger(os.path.join(self.logger_folder, self.ref_test_name +"_" + time.strftime("%Y-%m-%d_%H:%M:%S")+ LOG_EXTENSION))
+        self.log_params()
+
+    def log_params(self):
+        self.log.write("parameters", "Test name: " + self.ref_test_name)
+        self.log.write("parameters", "Test time: " + str(self.ref_timestamp))
+        self.log.write("parameters", "Logger folder: " + self.logger_folder)
+
+
+    def check_create_server(self):
+
+        self.log.write("info", "creating servers..")
+        for user in self.users:
+            global r
+            try:
+                r = self.session.create_server(self.port, user, self.data, self.verify)
```

Comment:
Why do you have a helper class for doing this if you only call it here? Or is there any other place?

Reply:
The API is being called only in this test file ( In earlier implementation it was called many times), I would remove it from there. 
Quoting from previous comment
 > Since API calls is also replicated in many tests, I would create a parent class called JupyterhubTest (child of Test - the class that implements the logs and all), which implements everything required to call the Jupyterhub API.
I'm little confused here. Should the class JupyterhubAPI contain only those API calls which are being replicated?

Reply:
What I mean if for you to create an aux function for performing calls to the API. And then you use it by passing the endpoint where you want to collect the info and some extra params.

Reply:
Got it!